encoding/codecv2/codecv2.go

```diff
@@ -0,0 +1,622 @@
+		BlobVersionedHash:      common.BytesToHash(data[57:89]),
+		ParentBatchHash:        common.BytesToHash(data[89:121]),
+		SkippedL1MessageBitmap: data[121:],
+	}
+
+	return b, nil
+}
+
+// Encode serializes the DABatch into bytes.
+func (b *DABatch) Encode() []byte {
+	batchBytes := make([]byte, 121+len(b.SkippedL1MessageBitmap))
+	batchBytes[0] = b.Version
+	binary.BigEndian.PutUint64(batchBytes[1:], b.BatchIndex)
+	binary.BigEndian.PutUint64(batchBytes[9:], b.L1MessagePopped)
+	binary.BigEndian.PutUint64(batchBytes[17:], b.TotalL1MessagePopped)
+	copy(batchBytes[25:], b.DataHash[:])
+	copy(batchBytes[57:], b.BlobVersionedHash[:])
+	copy(batchBytes[89:], b.ParentBatchHash[:])
+	copy(batchBytes[121:], b.SkippedL1MessageBitmap[:])
+	return batchBytes
+}
+
+// Hash computes the hash of the serialized DABatch.
+func (b *DABatch) Hash() common.Hash {
+	bytes := b.Encode()
+	return crypto.Keccak256Hash(bytes)
+}
+
+// BlobDataProof computes the abi-encoded blob verification data.
+func (b *DABatch) BlobDataProof() ([]byte, error) {
+	if b.blob == nil {
+		return nil, errors.New("called BlobDataProof with empty blob")
+	}
+	if b.z == nil {
+		return nil, errors.New("called BlobDataProof with empty z")
+	}
+
+	commitment, err := kzg4844.BlobToCommitment(b.blob)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create blob commitment")
+	}
+
+	proof, y, err := kzg4844.ComputeProof(b.blob, *b.z)
+	if err != nil {
+		log.Crit("failed to create KZG proof at point", "err", err, "z", hex.EncodeToString(b.z[:]))
+	}
+
+	// Memory layout of ``_blobDataProof``:
+	// | z       | y       | kzg_commitment | kzg_proof |
+	// |---------|---------|----------------|-----------|
+	// | bytes32 | bytes32 | bytes48        | bytes48   |
+
+	values := []interface{}{*b.z, y, commitment, proof}
+	return BlobDataProofArgs.Pack(values...)
+}
+
+// Blob returns the blob of the batch.
+func (b *DABatch) Blob() *kzg4844.Blob {
+	return b.blob
+}
+
+// DecodeFromCalldata attempts to decode a DABatch and an array of DAChunks from the provided calldata byte slice.
+func DecodeFromCalldata(data []byte) (*DABatch, []*DAChunk, error) {
+	// TODO: implement this function.
+	return nil, nil, nil
+}
+
+// EstimateChunkL1CommitBlobSize estimates the size of the L1 commit blob for a single chunk.
+func EstimateChunkL1CommitBlobSize(c *encoding.Chunk) (uint64, error) {
+	batchBytes, err := constructBatchPayload([]*encoding.Chunk{c})
+	if err != nil {
+		return 0, err
+	}
+	blobBytes, err := compressScrollBatchBytes(batchBytes)
+	if err != nil {
+		return 0, err
+	}
+	return calculatePaddedBlobSize(uint64(len(blobBytes))), nil
+}
+
+// EstimateBatchL1CommitBlobSize estimates the total size of the L1 commit blob for a batch.
+func EstimateBatchL1CommitBlobSize(b *encoding.Batch) (uint64, error) {
+	batchBytes, err := constructBatchPayload(b.Chunks)
+	if err != nil {
+		return 0, err
+	}
+	blobBytes, err := compressScrollBatchBytes(batchBytes)
```

Comment:
Will it be expensive to call this many times repeatedly?

Reply:
No idea in mind, it seems impossible to do any incremental estimation when we use compression. If this is not a concern in your tests, then we don't need to worry about it. But maybe we can add some metrics to track the runtime in the monorepo, just in case.

Reply:
Yes. will add some in the monorepo.